src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
+    swap(lhs.agg_ext_sigma_, rhs.agg_ext_sigma_);
+
+    swap(lhs.col_marker_, rhs.col_marker_);
+
+    swap(lhs.B_potential_, rhs.B_potential_);
+    swap(lhs.D_trace_sum_, rhs.D_trace_sum_);
+    swap(lhs.D_trace_, rhs.D_trace_);
+    swap(lhs.F_potential_, rhs.F_potential_);
 }
 
-void GraphCoarsen::BuildPVertices(
-    std::vector<mfem::DenseMatrix>& vertex_target,
-    mfem::SparseMatrix& Pvertices, bool build_coarse_relation)
+void GraphCoarsen::ComputeVertexTargets(const ParMatrix& M_ext_global,
+                                        const ParMatrix& D_ext_global)
 {
-    // simply put vertex_target into Pvertices
+    const SparseMatrix& M_ext = M_ext_global.GetDiag();
+    const SparseMatrix& D_ext = D_ext_global.GetDiag();
+
+    int num_aggs = gt_.agg_ext_edge_.Rows();
 
-    const unsigned int nAggs = vertex_target.size();
-    const mfem::SparseMatrix& Agg_vertex(graph_topology_.Agg_vertex_);
-    int nvertices = Agg_vertex.Width();
-    int nlocal_fine_dofs, nlocal_coarse_dofs;
-    mfem::Array<int> local_fine_dofs;
+    DenseMatrix evects;
+    LocalEigenSolver eigs(max_evects_, spect_tol_);
 
-    int* Pvertices_i = new int[nvertices + 1];
-    Pvertices_i[0] = 0;
-    for (unsigned int i = 0; i < nAggs; ++i)
+    for (int agg = 0; agg < num_aggs; ++agg)
     {
-        GetTableRow(Agg_vertex, i, local_fine_dofs);
-        nlocal_coarse_dofs = vertex_target[i].Width();
-        nlocal_fine_dofs = local_fine_dofs.Size();
-        for (int j = 0; j < nlocal_fine_dofs; ++j)
-            Pvertices_i[local_fine_dofs[j] + 1] = nlocal_coarse_dofs;
+        std::vector<int> edge_dofs_ext = GetExtDofs(gt_.agg_ext_edge_, agg);
+        std::vector<int> vertex_dofs_ext = GetExtDofs(gt_.agg_ext_vertex_, agg);
+        std::vector<int> vertex_dofs_local = gt_.agg_vertex_local_.GetIndices(agg);
+
+        if (edge_dofs_ext.size() == 0)
+        {
+            vertex_targets_[agg] = DenseMatrix(1, 1, {1.0});
+            continue;
+        }
+
+        SparseMatrix M_sub = M_ext.GetSubMatrix(edge_dofs_ext, edge_dofs_ext, col_marker_);
+        SparseMatrix D_sub = D_ext.GetSubMatrix(vertex_dofs_ext, edge_dofs_ext, col_marker_);
+        SparseMatrix D_sub_T = D_sub.Transpose();
+
+        D_sub_T.InverseScaleRows(M_sub);
+
+        SparseMatrix DMinvDT = D_sub.Mult(D_sub_T);
+
+        eigs.Compute(DMinvDT, evects);
+
+        if (evects.Cols() > 1)
+        {
+            DenseMatrix evects_ortho = evects.GetCol(1, evects.Cols());
+            agg_ext_sigma_[agg] = D_sub_T.Mult(evects_ortho);
+        }
+        else
+        {
+            agg_ext_sigma_[agg].SetSize(D_sub_T.Rows(), 0);
+        }
+
+        DenseMatrix evects_restricted = RestrictLocal(evects, col_marker_,
+                                                      vertex_dofs_ext, vertex_dofs_local);
+
+        VectorView first_vect = evects_restricted.GetColView(0);
+        vertex_targets_[agg] = smoothg::Orthogonalize(evects_restricted, first_vect, 1, max_evects_);
```

Comment:
is the scope `smoothg` necessary here?

Reply:
At one point there was a similarly named function in the class, but it is now no longer necessary. :+1: